Bluetooth: controller: Reduce sys_le16_to_cpu re-conversions

cvinayak opened the pull request:
Remove the use of sys_le16_to_cpu on HCI command parameter,
use auto-variables instead to remember le16 values instead.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Reduce sys_le16_to_cpu re-conversions

Remove the use of sys_le16_to_cpu on HCI command parameter,
use auto-variables instead to remember le16 values instead.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix vs_write_tx_power_level for overlapping mem

Fix HCI Vendor Specific Write Tx Power Level command for
overlapping memory of command and response buffer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

asmk-ot commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1542,74 +1547,79 @@ static void le_setup_iso_path(struct net_buf *buf, struct net_buf **evt)
 
 	rp = hci_cmd_complete(evt, sizeof(*rp));
 	rp->status = status;
-	rp->handle = cmd->handle;
```
No reconversion going on. Why not leave it at cmd->handle? Is there a risk of cmd->.. being overwritten?

cvinayak replied:
refer to https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L9243

Its been longtime... things may have changed... but I believe, the command buffer is reused to send back command complete

cvinayak replied:
sorry, I am trying to make different commit, for reducing the conversion, fixing the overwriting etc... lately, merged contribution have this inconsistency of overwrites and missing conversions.... keeping this PR in draft till I get the different commits correct

cvinayak replied:
practically the overwrites may not be noticed, but the correct needs to be correct so that in the future other parameters returns should not get overwritten.

wopu-ot replied:
I think fixing the overwrites is way more important than eliminating some endianness conversions (which will be NOPs for most architectures anyway). I suspect that a reasonably smart compiler will eliminate the conversion/re-conversion, but I need to check whether this is actually the case.

cvinayak replied:
As I am developing on little-endian machines, I did not see any conversion code in my builds before and after the change in this PR (flash usage remained same, as expected). Please check in big-endian machine, if the PR does not increase code size compared to before the PR.

wopu-ot replied:
I quick check did not reveal a code size difference between the two patterns below. I think the first pattern is more readable, but I will not object if other reviewers find the second pattern preferrable.
```
handle = sys_le16_to_cpu(cmd->handle)
...
rp->handle = sys_cpu_to_le16(handle)
```

```
handle_le16 = cmd->handle;
handle = sys_le16_to_cpu(handle_le16)
...
rp->handle = handle_le16
```

cvinayak replied:
I have done an experiment to use `sys_be16_to_cpu` on a little-endian machine and compared the assembly instructions:

![image](https://user-images.githubusercontent.com/6350656/100569967-df3f0380-32cf-11eb-8b91-90bae67e2008.png)

![image](https://user-images.githubusercontent.com/6350656/100569780-5d4eda80-32cf-11eb-82b8-4980e0d3d07e.png)

There is clearly added instructions on re-conversions in the first pattern compared to the use of auto-variable (register r4) to remember the value before conversion. (I will assume this could be observed when using `sys_le16_to_cpu` on a big-endian machine. And of course, this could be compiler dependent when optimizations is used).

asmk-ot commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1542,74 +1547,79 @@ static void le_setup_iso_path(struct net_buf *buf, struct net_buf **evt)
 
 	rp = hci_cmd_complete(evt, sizeof(*rp));
 	rp->status = status;
-	rp->handle = cmd->handle;
```
same here and other places

cvinayak committed:
Bluetooth: controller: Fix remember handle to be returned

Fix use of cmd->handle as return handle, as cmd buffer may
be reused as cmd complete buffer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

wopu-ot reviewed (approved):
LGTM

cvinayak commented:
@carlescufi or @kruithofa need yours as second approval here

sjanc commented:
I'm bit confused... what is the exact purpose of this patch?

cvinayak commented:
> 
> 
> I'm bit confused... what is the exact purpose of this patch?

Purpose is to not access cmd buffer after the same buffer is being used to generate the cmd response/status event. Like fixes in this commit af0524d3fc5338c611317bfa8d9c47ea9bfb5a33

sjanc commented:
OK, than this should be mentioned in commit message

FWIW I also find following pattern more readable:
handle = sys_le16_to_cpu(cmd->handle)
...
rp->handle = sys_cpu_to_le16(handle)

kruithofa commented:
I agree with Szymon (and Wolfgang) that the following is more readable:

handle = sys_le16_to_cpu(cmd->handle)
...
rp->handle = sys_cpu_to_le16(handle)

the difference in code-size is negligible, so I vote for readability over the minor code-size increase

cvinayak commented:
There are atleast 23 instances in the file (even more to come in the future), 8 bytes extra for conversion if the machine is big-endian, or 184 bytes. But is not for LE machines though.

I do not understand how `handle_le16` is very unreadable compared to `sys_cpu_to_le16(handle)`.

cvinayak commented:
@thoh-ot what is your call on the changes here?

thoh-ot commented:
The compiler we use for our big endian architecture, sees no change in program memory between the two patterns so my verdict would be to drop this PR.
